torch/_subclasses/fake_tensor.py

```diff
@@ -1326,6 +1495,7 @@ def __exit__(self, a, b, c):
             if maybe_prev_fake_mode is not None:
                 torch._C._set_dispatch_mode(maybe_prev_fake_mode)
 
+    @cache_dispatch
     def dispatch(self, func, types, args=(), kwargs=None):
```

Comment:
How does this interact with aot autograd? I thought we used `FakeTensor` to trace through aten dispatch, so by caching do we not miss nodes in the graph?

Reply:
Good question. @ezyang @Chillee What do you think?

Reply:
FakeTensor lives after ProxyTensor in the mode dispatch hierarchy.

For example, if you wanted to decompose `addcmul` into `add` and `mul`, the ordering would look like

```
ProxyTensor: addcmul
=> (hits decomposition table, decompose into add and mul)
ProxyTensor: add
=> (redispatch)
FakeTensor: add <----- this is where we're interceding for the cache.
```

Reply:
Do we register `CompositeImplicitAutograd` implementations as decompositions, such that  `FakeTensor.__torch_dispatch__` is never called for `CompositeImplicitAutograd` operators?

Reply:
Actually thinking again, `CompositeImplicitAutograd` come before `Python` in the dispatch priority so there is no issue here.

Reply:
Decompositions at the `CompositeImplicitAutograd` level are actually somewhat of a different mechanism, since it doesn't even go through the `Python` dispatch key at all.

But either way, the important thing is that it all happens *before* the FakeTensor dispatch, so you don't need to worry about it.

Put another way, FakeTensor dispatch will never "see" a ProxyTensor.